Implement custom helpers

juanazam opened the pull request:
Implements #22

juanazam commented:
@san650 I have finished with the discussed suggestions, please check it out :D

san650 commented on test-support/page-object/custom-helper.js:
```diff
@@ -0,0 +1,30 @@
+                         key,
+                         parent);
+          } else if ($.isFunction(customHelperRes)) {
+            return customHelperRes(scopedSelector);
```
I think we don't need to pass the scope as an argument here. We just need to invoke the function, right?

juanazam replied:
Following the spec we agreed on, we must pass the result of calling `qualifySelector(parent.scope, selector)` to get the parent's scope when the user defined function is called.

```
var input = customHelper(function(selector, options) {
  let hasError = $(selector).parent().hasClass('has-error');

  return {
    value: value(selector),
    hasError: function() {
      return hasError;
    }
  };
});

var page = PageObject.build({
  scope: 'form',
  userName: input('#userName')
});
```
In this case, the input function should receive `form #userName` as a parameter.

san650 commented on test-support/page-object/custom-helper.js:
```diff
@@ -0,0 +1,30 @@
+          if ($.isPlainObject(customHelperRes)) {
+            return build($.extend(customHelperRes, { scope: scopedSelector }),
+                         key,
+                         parent);
```
I think the order of the parameters in the `extend` is not the best.

If the customHelperRes has a `scope: 'something'` attribute, we're overriding it with our `{ scope: scopedSelector }` and may not be the desired effect.

I think the order could be `$.extend({ scope: scopedSelector }, customHelperRes)`, this way the user can override the scope.

juanazam replied:
:+1:

san650 commented on test-support/page-object/custom-helper.js:
```diff
@@ -0,0 +1,30 @@
+          // eval
+          let scopedSelector = qualifySelector(parent.scope, selector);
+
+          let customHelperRes = userDefinedFunction(scopedSelector, options);
```
Maybe we can simplify the code by renaming some variables `s/customHelperRes/response` `s/scopedSelector/qualifiedSelector/`

Also, maybe we can improve readability by having only one return statement

```js
if ($.isPlainObject(response)) {
    processedResponse = build(...);
} else if ($.isFunction(response)) {
    processedResponse = response();
} else {
    processedResponse = response;
}

return processedResponse;
```

What do you think?

juanazam replied:
:+1:

san650 commented on DOCUMENTATION.md:
```diff
@@ -615,6 +616,87 @@ var page = PO.build({
+Defines a way in which custom components can be created using
+information of their surrounding context.
+
+It expects a function with the following signature:
```
If you look at the other attributes, we add a _attribute signature_ section after the short description.

Something like...

```markdown
Attribute signature

` ``js
PO.customHelper(userFunction);
` ``

Where the provided user function has the following signature

` ``js
function(selector, options) {
  // user __code__ goes here
  return Value;
}
` ``

...

```

Or something on the line

juanazam replied:
:+1:

san650 commented on DOCUMENTATION.md:
```diff
@@ -615,6 +616,87 @@ var page = PO.build({
+  myText: {
+    scope: '#myInput',
+    isAdminField: admin(),
+    isDisabled: disabled()
```
Can we change `disabled` and `admin` function names to be `isDisabled` and `isAdmin`? So we follow the same convention as the rest of the project

san650 commented:
@juanazam This is great! Sorry I did a lot of comments but I think they're just small adjustments.

I'm going to test the functionality on a real project now and let you know how it goes :smile:

juanazam commented:
@san650 I addressed all issues except the discussion around if the selector should be passed as a parameter when the user defined function returns a function. :smile:
